solidity/contracts/converter/BancorConverter.sol

```diff
@@ -554,26 +569,27 @@ contract BancorConverter is IBancorConverter, SmartTokenController, ContractRegi
       * @param _reserveToken    reserve token contract address
       * @param _depositAmount   amount of tokens to deposit (in units of the reserve token)
       * @param _minReturn       if the conversion results in an amount smaller than the minimum return - it is cancelled, must be nonzero
+      * @param _beneficiary     wallet to receive the conversion result
       * 
       * @return amount of tokens received (in units of the smart token)
     */
-    function buy(IERC20Token _reserveToken, uint256 _depositAmount, uint256 _minReturn) internal returns (uint256) {
-        (uint256 amount, uint256 feeAmount) = getPurchaseReturn(_reserveToken, _depositAmount, msg.value);
+    function buy(IERC20Token _reserveToken, uint256 _depositAmount, uint256 _minReturn, address _beneficiary) internal returns (uint256) {
+        (uint256 amount, uint256 feeAmount) = getPurchaseReturn(_reserveToken, _depositAmount);
 
         // ensure the trade gives something in return and meets the minimum requested amount
         require(amount != 0 && amount >= _minReturn);
 
-        // transfer funds from the caller in the reserve token
-        if (_reserveToken == IERC20Token(0)) {
+        // ensure that the input amount was already deposited
+        if (_reserveToken == IERC20Token(0))
             require(msg.value == _depositAmount);
-        }
-        else {
-            require(msg.value == 0);
-            ensureTransferFrom(_reserveToken, msg.sender, this, _depositAmount);
-        }
+        else
+            require(msg.value == 0 && _reserveToken.balanceOf(this).sub(getReserveBalance(_reserveToken)) >= _depositAmount);
 
-        // issue new funds to the caller in the smart token
-        token.issue(msg.sender, amount);
+        // sync the reserve balance
+        syncReserveBalance(_reserveToken);

```

Comment:
May as well embed the code of function `syncReserveBalance` into the if/else clause above.

Reply:
What is the benefit that's worth to skip code reuse here?